From 448066353e87bfe64c104ad3397a1e0c07789ef7 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 1 Oct 2024 10:59:36 +0800
Subject: [PATCH 221/265] mtk: mt76: mt7996: return -EINVAL on not getting
 band_idx for specific vendor commands

Some vendor commands need the band_idx, while others do not.
Returning -EINVAL when unsuccessfully getting band_idx for those do not
need a band_idx is unnecessary.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt7996/vendor.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/mt7996/vendor.c b/mt7996/vendor.c
index 39a5d714..ece8fabf 100644
--- a/mt7996/vendor.c
+++ b/mt7996/vendor.c
@@ -1031,6 +1031,7 @@ static int mt7996_vendor_wireless_ctrl(struct wiphy *wiphy,
 	u8 val8, band_idx, link_id = 0;
 	u16 val16;
 	u32 val32;
+	bool band_idx_get = false;
 
 	err = nla_parse(tb, MTK_VENDOR_ATTR_WIRELESS_CTRL_MAX, data, data_len,
 			wireless_ctrl_policy, NULL);
@@ -1046,14 +1047,18 @@ static int mt7996_vendor_wireless_ctrl(struct wiphy *wiphy,
 
 	rcu_read_lock();
 	mconf = rcu_dereference(mvif->link[link_id]);
-	if (!mconf || !mconf->phy) {
-		rcu_read_unlock();
-		return -EINVAL;
+	if (mconf && mconf->phy) {
+		band_idx = mconf->phy->mt76->band_idx;
+		band_idx_get = true;
 	}
-
-	band_idx = mconf->phy->mt76->band_idx;
 	rcu_read_unlock();
 
+	if (!band_idx_get && (tb[MTK_VENDOR_ATTR_WIRELESS_CTRL_OFDMA] ||
+	    tb[MTK_VENDOR_ATTR_WIRELESS_CTRL_NUSERS_OFDMA] ||
+	    tb[MTK_VENDOR_ATTR_WIRELESS_CTRL_MIMO] ||
+	    tb[MTK_VENDOR_ATTR_WIRELESS_CTRL_CERT]))
+		return -EINVAL;
+
 	if (tb[MTK_VENDOR_ATTR_WIRELESS_CTRL_OFDMA]) {
 		val8 = nla_get_u8(tb[MTK_VENDOR_ATTR_WIRELESS_CTRL_OFDMA]);
 		val32 = FIELD_PREP(RATE_CFG_MODE, RATE_PARAM_FIXED_OFDMA) |
-- 
2.45.2

